backend/internal/system/immutable_resource/composite_helpers.go

```diff
@@ -0,0 +1,418 @@
+/*
+ * Copyright (c) 2025, WSO2 LLC. (https://www.wso2.com).
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package immutableresource
+
+import (
+	"errors"
+	"fmt"
+)
+
+// CompositeGetHelper implements the dual-store Get pattern for composite stores.
+//
+// Type Parameter:
+//   - T: The domain entity type (e.g., OrganizationUnit, Application, IDPDTO)
+//
+// Parameters:
+//   - id: Resource identifier to retrieve
+//   - dbGetter: Function that retrieves from database store
+//   - fileGetter: Function that retrieves from file store
+//   - notFoundError: Domain-specific "not found" error to return
+//
+// Returns:
+//   - Entity of type T
+//   - Error if not found in either store or other errors occur
+//
+// Behavior:
+//  1. Try database store first (runtime/mutable resources)
+//  2. If found, return it
+//  3. If not found, try file store (YAML/immutable resources)
+//  4. If found in file, return it
+//  5. If not found in either, return notFoundError
+func CompositeGetHelper[T any](
+	id string,
+	dbGetter func(string) (T, error),
+	fileGetter func(string) (T, error),
+	notFoundError error,
+) (T, error) {
+	var zero T // Zero value for type T
+
+	// Phase 1: Check database store (mutable resources)
+	entity, err := dbGetter(id)
+	if err == nil {
+		return entity, nil
+	}
+
+	// Phase 2: If not found in DB, check file store (immutable resources)
+	if errors.Is(err, notFoundError) {
+		entity, fileErr := fileGetter(id)
+		if fileErr == nil {
+			return entity, nil
+		}
+		// Not found in either store
+		return zero, notFoundError
+	}
+
+	// Database error (not "not found")
+	return zero, err
+}
+
+// CompositeGetByFieldHelper implements dual-store retrieval by any field for composite stores.
+//
+// Use cases:
+//   - GetOrganizationUnitByPath(handles []string)
+//   - GetApplicationByName(name string)
+//   - GetIDPByName(name string)
+//
+// Type Parameter:
+//   - T: The domain entity type
+//
+// Parameters:
+//   - dbGetter: Function that retrieves from database store
+//   - fileGetter: Function that retrieves from file store
+//   - notFoundError: Domain-specific "not found" error
+//
+// Returns:
+//   - Entity of type T
+//   - Error if not found in either store
+func CompositeGetByFieldHelper[T any](
+	dbGetter func() (T, error),
+	fileGetter func() (T, error),
+	notFoundError error,
+) (T, error) {
```

Comment:
Do we need seperate function for this. can't we improve the, CompositeGetHelper to make it generic to support get by string
```
func CompositeGetHelper[T any](
	dbGetter func() (T, error),
	fileGetter func() (T, error),
	notFoundError error,
) (T, error) {
```